backends/bmv2/common/expression.cpp

```diff
@@ -184,32 +184,57 @@ void ExpressionConverter::postorder(const IR::Constant* expression)  {
 
 void ExpressionConverter::postorder(const IR::ArrayIndex* expression)  {
     auto result = new Util::JsonObject();
-    result->emplace("type", "header");
     cstring elementAccess;
 
+    LOG2("left: " << expression->left << " right: " << expression->right);
     // This is can be either a header, which is part of the "headers" parameter
     // or a temporary array.
-    if (expression->left->is<IR::Member>()) {
+    if (auto mem = expression->left->to<IR::Member>()) {
         // This is a header part of the parameters
-        auto mem = expression->left->to<IR::Member>();
         auto parentType = typeMap->getType(mem->expr, true);
         BUG_CHECK(parentType->is<IR::Type_StructLike>(), "%1%: expected a struct", parentType);
         auto st = parentType->to<IR::Type_StructLike>();
         auto field = st->getField(mem->member);
         elementAccess = field->controlPlaneName();
-    } else if (expression->left->is<IR::PathExpression>()) {
+    } else if (auto path = expression->left->to<IR::PathExpression>()) {
         // This is a temporary variable with type stack.
-        auto path = expression->left->to<IR::PathExpression>();
         elementAccess = path->path->name;
     }
 
     if (!expression->right->is<IR::Constant>()) {
-        ::error(ErrorType::ERR_INVALID, "all array indexes must be constant", expression->right);
+        result->emplace("type", "expression");
+        const IR::Expression* ex = expression->right;
+        std::vector<cstring> cv;
+        while (auto rmem = ex->to<IR::Member>()) {
```

Comment:
Good catch.  Let me see what I can do.

Reply:
It's not a visit subexpression and recursive call issue. 
Arithmetic expression such as `a+b` for constat ArrayIndex is not processed in the backend.  

I decided to debug a constant array index `a+b`and see that the expression is simplified to the sum of `a+b` by the time I dump FrontEndLast.  

`hdr.vector[0+1].e = hdr.pool[1].val + 1;`

changes to

`hdr.vector[1].e = hdr.pool[1].val + 1;`

Since a runtime index cannot be constant folded, the ArrayIndex reaches the bmv2 backend as `hdr.pool[hdr.ml.idx + 8s1].val`  and is processed first by 'void ExpressionConverter::postorder(const IR::IntMod* expression)' before processing calls `void ExpressionConverter::postorder(const IR::ArrayIndex* expression)`.  

With `a+b` for runtime index an IntMod expression is sent to postorder(ArrayIndex).  With this new code added

```
if (!expression->right->is<IR::Constant>()) {
        result->emplace("type", "expression");
        const IR::Expression* ex = expression->right;
        std::vector<cstring> cv;
        if (auto im = ex->to<IR::IntMod>()) { <=== start new code
	  if (auto bio = im->expr->to<IR::Add>()) {
	    ex = bio->left;
	    std::cout << "AIM:" << bio->left << std::endl;
	  }
        }  <=== end new code.
```

bio->left is `hdr.ml.idx;` and rest of the code runs to completion. 

A question for Antonin:

If the P4 code is ` hdr.pool[hdr.ml.idx+1]`, what should the JSON look like because we are in uncharted territory.  Would it be the one shown below?

"value" : ["ml", "idx+1"]

Reply:
I am almost certain that is not what should be in the BMv2 JSON.

There is already defined JSON expressions for expressions like (hdr.ml.idx + 1) appearing on the right hand side of an assignment.

Most likely whatever that JSON is, is what should be in place of the field member access hdr.ml.idx now.

Reply:
FYI, attached is a proposed second test P4 program and STF file that exercises arithmetic expressions inside of the run time array indexes.  I have not tested whether the STF is correct or not, since there is no implementation to try it on, but there are extensive comments in the STF file explaining why I think it might be correct.

Reply:
@jafingerhut I know.  However, the postorder(ArrayIndex) method does not deal with arithmetic expressions.  Some other code does.   I am not sure right now, if I add arithmetic to the runtime index processing in postorder(ArrayIndex).

Reply:
Arithmetic expressions with runtime index is not supported by this PR code changes yet.

Reply:
I understand it is not supported yet.  If/when it is supported, it will be useful to have a test case to exercise that it works.  Hence my second attached zip file.